server/fleet/errors.go

```diff
@@ -267,3 +271,71 @@ func NewErrorf(code int, format string, args ...interface{}) error {
 func (ge *Error) Error() string {
 	return ge.Message
 }
+
+// UserMessageError is an error that adds the UserMessage interface
+// implementation.
+type UserMessageError struct {
+	error
+	statusCode int
+}
+
+// NewUserMessageError creates a UserMessageError that will translate the
+// error message of err to a user-friendly form. If statusCode is > 0, it
+// will be used as the HTTP status code for the error, otherwise it defaults
+// to http.StatusUnprocessableEntity (422).
+func NewUserMessageError(err error, statusCode int) *UserMessageError {
+	return &UserMessageError{err, statusCode}
+}
+
+var rxJSONUnknownField = regexp.MustCompile(`^json: unknown field "(.+)"$`)
+
+// UserMessage implements the user-friendly translation of the error if its
+// root cause is one of the supported types, otherwise it returns the error
+// message.
+func (e UserMessageError) UserMessage() string {
+	cause := Cause(e.error)
```

Comment:
Hmm I guess it could make sense in `NewUserMessage` to prevent such a value from being created with a nil error. When asking for the `UserMessage`, it seems a bit late to me, we expect this to be an error at that point.

Reply:
I added a check in `NewUserMessageError` so that it returns nil if the provided error is nil, let me know how you feel about that!

Reply:
that looks super natural, thanks!